torchtune/data/_messages.py

```diff
@@ -726,14 +772,94 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
             Message(
                 role="user",
                 content=prompt,
-                masked=not self.train_on_input,
                 eot=True,
             ),
             Message(
                 role="assistant",
                 content=sample[key_output],
-                masked=False,
                 eot=True,
             ),
         ]
+        mask_messages(messages, self.masking_strategy)
         return {"messages": messages}
+
+
+def validate_messages(
+    messages: List[Message],
+) -> None:
+    """
+    Given a list of messages, ensure that messages form a valid
+    back-and-forth conversation. An error will be raised if:
+
+    - There is a system message that's not the first message
+    - There are two consecutive user messages
+    - An assistant message comes before the first user message
+    - The message is empty
+    - Messages are shorter than length of 2 (min. one user-assistant turn)
+
+
+    Args:
+        messages (List[Message]): the messages to validate.
+
+    Raises:
+        ValueError: If the messages are invalid.
+    """
+    if len(messages) < 2:
+        raise ValueError(
+            f"Messages must be at least length 2, but got {len(messages)} messages"
+        )
+
+    last_turn = "assistant"
+    for i, message in enumerate(messages):
+        if message.role == "assistant" and last_turn != "user":
+            raise ValueError(
+                f"Assistant message before expected user message at index {i} in messages"
+            )
+        if message.role == "user" and last_turn == "user":
+            raise ValueError(
+                f"Two consecutive user messages at index {i} and {i - 1} in messages"
+            )
+        if message.role == "system" and i > 0:
+            raise ValueError(
+                f"System message at index {i} in messages, but system messages must come first"
+            )
+        last_turn = message.role
+
+
+def mask_messages(messages: List[Message], masking_strategy: str) -> None:
+    """
+    Set the masked attribute for each message in the list based on the specified masking strategy.
+
+    Args:
+        messages (List[Message]): a list of messages to mask.
+        masking_strategy (str): masking strategy to use.
+            Must be one of: `train_on_all`, `train_on_assistant`, `train_on_last`.
+            - ``train_on_all``: both user and assistant messages are unmasked
+            - ``train_on_assistant``: user messages are masked, only assistant messages are unmasked
+            - ``train_on_last``: only the last assistant message is unmasked
+
+    Raises:
+        ValueError: If the masking strategy is not one of the supported strategies:
+                    `train_on_all`, `train_on_assistant`, `train_on_last`.
+    """
+    if masking_strategy not in ["train_on_all", "train_on_assistant", "train_on_last"]:
```

Comment:
One suggestion that I didn't mention in the issue is that we could make `masking_strategy` an Enum so that it can be type checked more easily. users can still specify as a string in the config and in the message transforms above, but here in `mask_messages` you could convert to the Enum, and the ValueError will automatically be captured.

Reply:
Converted masking_strategy to Enum and raises ValueErroe automatically if doesn't match with mapping